lib/src/io.dart

```diff
@@ -893,110 +882,40 @@ final String _pathTo7zip = (() {
-        // preserving ownership anyway, so we just set them to "pub".
-        // TODO(rnystrom): This assumes contents does not contain any
-        // directories.
-        var mtreeHeader = '#mtree\n/set uname=pub gname=pub type=file\n';
-
-        // We need a newline at the end, otherwise the last file would get
-        // ignored.
-        stdin =
-            mtreeHeader + contents.join('\n').replaceAll(' ', r'\040') + '\n';
-      }
-
-      // Setting the working directory should be unnecessary since we pass an
-      // explicit base directory to tar. However, on Mac when using an mtree
-      // input file, relative paths in the mtree file are interpreted as
-      // relative to the current working directory, not the "--directory"
-      // argument.
-      var process = await startProcess(_tarPath, args, workingDir: baseDir);
-      process.stdin.add(utf8.encode(stdin));
-      process.stdin.close();
-      return process.stdout;
+  var buffer = StringBuffer();
+  buffer.write('Creating .tar.gz stream containing:\n');
+  contents.forEach(buffer.writeln);
+  log.fine(buffer.toString());
+
+  baseDir ??= path.current;
+  baseDir = path.absolute(baseDir);
+
+  final tarContents = Stream.fromIterable(contents.map((entry) {
+    entry = path.absolute(entry);
+    if (!path.isWithin(baseDir, entry)) {
+      throw ArgumentError('Entry $entry is not inside $baseDir.');
     }
 
-    // Don't use [withTempDir] here because we don't want to delete the temp
-    // directory until the returned stream has closed.
-    var tempDir = await _createSystemTempDir();
-
-    try {
-      // Create the file containing the list of files to compress.
-      var contentsPath = path.join(tempDir, 'files.txt');
-      writeTextFile(contentsPath, contents.join('\n'));
-
-      // Create the tar file.
-      var tarFile = path.join(tempDir, 'intermediate.tar');
-      var args = ['a', '-w$baseDir', tarFile, '@$contentsPath'];
-
-      // We're passing 'baseDir' both as '-w' and setting it as the working
-      // directory explicitly here intentionally. The former ensures that the
-      // files added to the archive have the correct relative path in the
-      // archive. The latter enables relative paths in the "-i" args to be
-      // resolved.
-      await runProcess(_pathTo7zip, args, workingDir: baseDir);
-
-      // GZIP it. 7zip doesn't support doing both as a single operation.
-      // Send the output to stdout.
-      args = ['a', 'unused', '-tgzip', '-so', tarFile];
-      return (await startProcess(_pathTo7zip, args))
-          .stdout
-          .transform(onDoneTransformer(() => deleteEntry(tempDir)));
-    } catch (_) {
-      deleteEntry(tempDir);
-      rethrow;
-    }
-  })));
+    final relative = path.relative(entry, from: baseDir);
+    // On Windows, we can't open some files without normalizing them
+    final file = File(path.normalize(entry));
+    final stat = file.statSync();
```

Comment:
Can we test if this is a symlink... If so maybe we should print a warning line that the symlink will be embedded as a file, and NOT a link.

IMO, it's probably preferable to keep the format simple and avoid fancy features like symlinks in distributed packages.

Reply:
I couldn't find any documentation stating whether it is acceptable to exclude directory entries or not.

GNU tar has explicit support for it though (it recovers from `ENOENT` by creating parent directories), and I'm not aware of any tar client that doesn't support this.